crates/shared/src/price_estimation/buffered.rs

```diff
@@ -0,0 +1,251 @@
+            let mut in_flight_requests = self.inner.in_flight_requests.lock().unwrap();
+
+            Inner::collect_garbage_with_lock(&mut in_flight_requests);
+
+            // For each `Query` either get an in-flight request or keep the `Query` to forward it to the
+            // inner price estimator.
+            let (active_requests, remaining_queries): (Vec<_>, Vec<_>) = queries
+                .iter()
+                .map(
+                    |query| match in_flight_requests.get(query).map(WeakShared::upgrade) {
+                        // NOTE: Technically it's possible under very specific circumstances
+                        // that the `active_request` is sitting in the cache for a long time
+                        // without making progress. If somebody else picks it up and polls
+                        // it to completion a timeout error will most likely be the result.
+                        // See https://github.com/gnosis/gp-v2-services/pull/1677#discussion_r813673692
+                        // for more details.
+                        Some(Some(active_request)) => (Some(active_request), None),
+                        _ => (None, Some(*query)),
+                    },
+                )
+                .unzip();
+
+            // Create future which estimates all `remaining_queries` in a single batch.
+            let fetch_remaining_estimates = {
+                let remaining_queries: Vec<_> =
+                    remaining_queries.iter().flatten().cloned().collect();
+                let inner = self.inner.clone();
+                async move { inner.estimator.estimates(&remaining_queries).await }
+                    .boxed()
+                    .shared()
+            };
+
+            // Create a `SharedEstimationRequest` for each individual `Query` of the batch. This
+            // makes it possible for a `batch_2` to await the queries which it is interested in of the
+            // in-flight `batch_1`. Even if the estimator which requested `batch_1` stops polling it, the
+            // estimator of `batch_2` can still poll `batch_1` to completion by polling the
+            // `SharedEstimationRequest` it is actually interested in.
+            let new_requests: Vec<_> = remaining_queries
+                .into_iter()
+                .flatten()
+                .enumerate()
+                .map(|(index, query)| {
+                    let fetch_remaining_estimates = fetch_remaining_estimates.clone();
+                    (
+                        query,
+                        async move { fetch_remaining_estimates.await[index].clone() }
+                            .boxed()
+                            .shared(),
+                    )
+                })
+                .collect();
+            in_flight_requests.extend(
+                new_requests
+                    .iter()
+                    .map(|(query, fut)| (*query, fut.downgrade().unwrap())),
```

Comment:
uber-nano-can-be-ignored-nit: When I read this line, I found myself needing to consult the docs to read up on why the `downgrade` could fail (future has already been polled to completion). Maybe this could be changed to an `expect` since understanding details of how `Shared` work is not common knowledge?

That being said, it took me 20 seconds at most to understand what was going on, so no strong feelings about this suggestion.

Reply:
At some point in time I must have forgotten to rebase this PR properly. The code you highlighted was already merged into main.
But I can sneak this suggestion into this PR, no problem.